Patch txpool core/list for managing multi currencies

hbandura opened the pull request:
Patches the txpool for multi currency. Validation, List (per user nonce heap), TxComparator(pricedHeap)

hbandura committed:
Add celo_list

CRU

Move comment line

Rename ToCurrencyValue to DenominateInCurrency

list modification for celo changes
list.Add and list.FilterBalance not working yet

Cherry-pick of 704201f98811bd47a7ad921f3f08c9cb5ce3527d (@palango fee currency management)

list.Add & some work on rates

Add tests for list totalcost & whitelist filter

Refactor FilterBalance for clarity & testiness

Add test for list FilterBalance (gaslimit pruning)

Add txcomparator

Use ratesComparator for pricedList

Add implementation of tx comparator

Implement GetRatesFromState

checking balance from txpool & lint fixes

Ignore l1cost for the moment

Add translate value impl

Remove 'areequaladdresses' since golang pointer comparisons work

Fix nil pointer

Change functions signature type

Change isCurrencyWhitelisted function to common

Fix comment

Add non-whitelisted test

Fix some tests failing
Changes list to use costCap in multicurrency fashion

Remove translator

Remove whitelist checker

Remove unnecessary struct

CRU

Move getbalance to celobackend

Add nonce replacement test to e2e tests

Remove feeCurrencyValidator

Move GetExchangeRates function to CeloBackend

Add log

Add current rates to legacypool

Add more test cases for nonce replacement bump

Code review updates

remove txComparator as type, alias as function

CRU

palango commented on core/txpool/legacypool/celo_legacypool.go:
```diff
@@ -0,0 +1,24 @@
+
+// filter Filters transactions from the given list, according to remaining balance (per currency, minus l1Cost)
+// and gasLimit. Returns drops and invalid txs.
+func (pool *LegacyPool) filter(list *list, addr common.Address, l1Cost *big.Int, gasLimit uint64) (types.Transactions, types.Transactions) {
```
You could also make this a member of `list`, as you're using many functions from it anyways.

hbandura replied:
I'll rewrite it like that and see how it looks

hbandura replied:
I didn't like how it looked by making it a member of the list, since I needed to pass some extra parameters from the pool.

palango commented on core/txpool/celo_rates.go:
```diff
@@ -23,11 +29,16 @@ func IsWhitelisted(exchangeRates common.ExchangeRates, feeCurrency *common.Addre
 	return ok
 }
 
+func TranslateValue(exchangeRates common.ExchangeRates, val1 *big.Int, currencyFrom *common.Address, currencyTo *common.Address) *big.Int {
```
We have that implemented already, if you only need to convert from celo to some currency: https://github.com/celo-org/op-geth/blob/715f16e2276b3e284b7b642e5ddfc9e39f7ce119/contracts/fee_currencies.go#L50

hbandura replied:
awesome, though the fn I meant is translating from one custom currency to another one, not passing through gold currency

palango commented on contracts/fee_currencies.go:
```diff
@@ -47,6 +47,17 @@ func GetBalanceOf(caller bind.ContractCaller, accountOwner common.Address, contr
 	return balance, nil
 }
 
+func ConvertCurrencyToGold(exchangeRates map[common.Address]*big.Rat, currencyAmount *big.Int, feeCurrency *common.Address) (*big.Int, error) {
```
Should use `ExchangeRates` type in the signature and below.
Also some simple tests would be nice, I should have added them as well.

hbandura replied:
I'll change it

palango commented on core/txpool/celo_rates.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+// IsWhitelisted checks if a given fee currency is whitelisted
+func IsWhitelisted(exchangeRates common.ExchangeRates, feeCurrency *common.Address) bool {
```
We also use this function in https://github.com/celo-org/op-geth/blob/b4ef923d8bc6318bfcadb8e6da5667e8d48c9967/core/vm/evm.go#L99

Would be nice to use one implementation.

hbandura replied:
I could move this to common. maybe and change all calls to these two functions to the new one

palango commented on core/txpool/celo_rates.go:
```diff
@@ -0,0 +1,117 @@
+	rateCache *lru.BasicLRU[common.Hash, common.ExchangeRates]
+)
+
+func GetRatesForState(config *params.ChainConfig, root common.Hash, state *state.StateDB) common.ExchangeRates {
```
Why is caching needed here? In my understanding the txpool always operates on the latest state and gets updated when a new head is available. Couldn't we trigger the exchange rate update on this event and get rid of the LRU here? Then we would also get rid of this extra function.

hbandura replied:
Yes, today I was self reviewing the PR and thought about eliminating the chain of fn calls with stateDB. I might try just holding the exchange rate in the txpool and see how it looks.

carterqw2 replied:
Have you checked this option in the end?

hbandura replied:
I somewhat did before by calling directly to that function. I improved this a bit today by removing the cache, and just keeping the currentRates as a member of the legacypool type, recalculating it when the currentState changes.

palango commented on core/txpool/legacypool/list.go:
```diff
@@ -281,9 +280,10 @@ type list struct {
 	strict bool       // Whether nonces are strictly continuous or not
 	txs    *sortedMap // Heap indexed sorted hash map of the transactions
 
-	costcap   *big.Int // Price of the highest costing transaction (reset only if exceeds balance)
```
Why don't we need `costcap` any more?

hbandura replied:
It's not that we don't need. costcap was an optimization where the highest gas price win the list would be the costcap. Then when adding a new tx, you can re-check the account balance, and to remove the txs with lower balance without actually checking them one by one, you can quickly check against the costcap and not do anything.

When I wrote it I thought that it's not that much important for us since checking the balances of the currencies is already orders of magnitude slower, but now that I think of it, I could still add it back and maybe it would look cleaner since I could already check beforehand with the costcaps which balances I have to check and send only that (right now it's being done on demand, and while it's an ok approach I feel like it's an extra type that could be avoided).

carterqw2 replied:
Feels like it could be a map like `totalcost`, I need to check if that would allow to simplify the code though.

hbandura replied:
Yes, I've already implemented it, but fixing some pesky bugs. The whole thing about mixing currencies and pointers to currencies has made it a little bit annoying, so I'm trying to keep some coherence in some parts

palango commented on core/txpool/legacypool/list.go:
```diff
@@ -308,11 +307,12 @@ func (l *list) Contains(nonce uint64) bool {
 	old := l.txs.Get(tx.Nonce())
 	if old != nil {
-		if old.GasFeeCapCmp(tx) >= 0 || old.GasTipCapCmp(tx) >= 0 {
+		// Short circuit for majority of problems
```
What are the "majority of problems"?

hbandura replied:
Haha, I must have been a bit asleep when I wrote that. I'll change it to 'short circuit when it's clear that the new tx is worse'

palango reviewed (commented):
Nice work, to me this feels like a better approach that the `celo_list`. Do you agree?

Some general notes:
- A short PR description is always helpful, with some hints on what might be surprising or some key design choices.
- Some tests are failing, I didn't check what causes them.
- There's still todos in the code, especially around error handling and the L1 cost. I didn't look into those.
- Are there other thing you intent on working in this PR?


One question around the design: There's lot's of new objects, and I'm not sure why you don't use functions instead. For example, the `whitelistcheck` struct is only created in `NewWhitelistChecker`, which is only used in `LegacyPool.filter`. Its `IsWhitelisted` method just passes down the request to `FeeCurrencyValidator`. For me this seems a bit meritless and I'd either use the validator directly or even the free `IsWhitelisted` function with the exchange rate stored in the txpool. What do you think?

hbandura commented:
> Nice work, to me this feels like a better approach that the `celo_list`. Do you agree?

Yes, but the celo_list approach had not be done to look cleaner, but to avoid the most common issues I used to have when merging upstream. However it might make sense to compromise a bit here.

 
> Some general notes:
> 
> * A short PR description is always helpful, with some hints on what might be surprising or some key design choices.
You are right, sorry about that.
> * Some tests are failing, I didn't check what causes them.

I'm fixing them on the go, I didn't want to wait to start the review, I doubt they will need a big change to be fixed (I can always be surprised)
> * There's still todos in the code, especially around error handling and the L1 cost. I didn't look into those.

Everything regarding L1Cost was intentional, I'm not sure if we are clear on how we want to manage that in the pool.
The logging I want to see how we are logging elsewhere first.
> * Are there other thing you intent on working in this PR?

No, unless there's something spotted in this review that I missed.
> 
> One question around the design: There's lot's of new objects, and I'm not sure why you don't use functions instead. For example, the `whitelistcheck` struct is only created in `NewWhitelistChecker`, which is only used in `LegacyPool.filter`. Its `IsWhitelisted` method just passes down the request to `FeeCurrencyValidator`. For me this seems a bit meritless and I'd either use the validator directly or even the free `IsWhitelisted` function with the exchange rate stored in the txpool. What do you think?

Many of this choices were made step by step and when you reach the final product you start cutting back, at least I already removed some as soon as I created the PR. 
The specific you mentioned, whitelistchecker, was because I didn't want the list to have the stateDb, so I fixed that in a different object and passed that.
Still, I have been thinking in remove the whole state hierarchy call and use an exchangerate fixed in the pool for mostly everything except balances. I will attempt to make some of those changes and see how it looks.

carterqw2 commented on core/txpool/celo_rates_test.go:
```diff
@@ -0,0 +1,144 @@
+			},
+			wantResult: 0,
+		}, {
+			name: "Different amounts of mixed currency 2",
+			args: args{
+				val1:         big.NewInt(100),
+				feeCurrency1: nil,
+				val2:         big.NewInt(45),
+				feeCurrency2: &currB,
+			},
+			wantResult: 0,
```
This test case is equivalent to the other one, might be worth changing the order of currencies in this context - make `feeCurrency1` `currB` and `feeCurrency2` `nil`.

hbandura replied:
changed it

carterqw2 commented on core/txpool/celo_rates_test.go:
```diff
@@ -0,0 +1,144 @@
+			},
+			wantResult: 1,
+		}, {
+			name: "Different amounts of same currency 2",
```
I would add test cases for two different currencies `currA` and `currB`

hbandura replied:
done!

carterqw2 commented on core/txpool/celo_rates_test.go:
```diff
@@ -0,0 +1,144 @@
+			got, err := CompareValue(exchangeRates, tt.args.val1, tt.args.feeCurrency1, tt.args.val2, tt.args.feeCurrency2)
+
+			if tt.wantErr && err == nil {
+				t.Error("Expected error in celoContextImpl.CompareFees")
```
Should we use `CompareValue` here, below and the name of the test then?

hbandura replied:
agreed. Changed it

carterqw2 commented on core/txpool/celo_rates.go:
```diff
@@ -0,0 +1,108 @@
+	} else {
+		exchangeRate2, ok = exchangeRates[*feeCurrency2]
+		if !ok {
+			return 0, fmt.Errorf("fee currency not registered: %s", feeCurrency1.Hex())
```
Should be `feeCurrency2.Hex()`

hbandura replied:
correct

carterqw2 commented on core/txpool/celo_rates.go:
```diff
@@ -0,0 +1,108 @@
+		}
+	}
+
+	if feeCurrency2 == nil {
+		exchangeRate2 = unitRate
+	} else {
+		exchangeRate2, ok = exchangeRates[*feeCurrency2]
+		if !ok {
+			return 0, fmt.Errorf("fee currency not registered: %s", feeCurrency1.Hex())
+		}
+	}
```
Might be worth extracting into a separate function as it's repeated twice to make the code more readable.

hbandura replied:
done

carterqw2 commented on core/txpool/celo_rates.go:
```diff
@@ -0,0 +1,108 @@
+	return leftSide.Cmp(rightSide), nil
+}
+
+func GetBalanceOf(config *params.ChainConfig, state *state.StateDB, account common.Address, feeCurrency *common.Address) *big.Int {
```
Feels like this method could live somewhere else. Could `CeloBackend` itself be a good place for that? @palango what do you think?

palango replied:
Yes, we can move it there, together with the `fee_currencies.go:GetBalanceOf` function.

hbandura replied:
I'm not quite sure of how you mean to move it, but I did this:
fee_currencies.go:GetBalanceOf is now CeloBackend.GetBalanceERC20(addr, addr), 
and the celo_rates GetBalanceOf was simply moved to the same file as celo_backend, but not being a member of the type.

carterqw2 commented on core/txpool/celo_validation.go:
```diff
@@ -15,29 +15,78 @@ var NonWhitelistedFeeCurrencyError = errors.New("Fee currency given is not white
+
+	// TranslateValue returns the given value expressed in a different currency, according
+	// to the exchange rate at the state tree presented.
+	TranslateValue(root common.Hash, st *state.StateDB, value *big.Int, fromCurrency *common.Address, toCurrency *common.Address) *big.Int
```
Maybe `ConvertCurrency` could be a better name, drawing inspiration from real life and the context of currency exchanges. Looks like the term currency conversion is quite wide-spread.

hbandura replied:
yes, I guess I wasn't inspired on the name. I'll change it

carterqw2 commented on core/txpool/celo_validation.go:
```diff
@@ -15,29 +15,78 @@ var NonWhitelistedFeeCurrencyError = errors.New("Fee currency given is not white
+	return TranslateValue(rates, value, fromCurrency, toCurrency)
+}
+
+type translator struct {
```
`translator` and `whitelistcheck` are identical and their methods are proxies to `fcv`/`FeeCurrencyValidator`. To me it's a sign that they belong to the same entity. Looks like we're using `TranslateValues`, `Balance` and `IsWhitelisted` methods from that entity, only `IsWhitelisted` can be considered to have anything to do with validation, so maybe a better naming is needed. It appears like a fee currencies helper of some sort. I would suggest to get rid of unnecessary interfaces and structures.

hbandura replied:
removed both translator and whitelistcheck, and tidied(?) up the method signatures

carterqw2 commented on core/celo_backend.go:
```diff
@@ -43,8 +47,34 @@ func (b *CeloBackend) CallContract(ctx context.Context, call ethereum.CallMsg, b
+	return balance, nil
+}
+
+func GetBalanceOf(config *params.ChainConfig, state *state.StateDB, account common.Address, feeCurrency *common.Address) *big.Int {
```
Looks like `FeeCurrencyValidator` or `feeval` struct to be precise is the only caller of this method. Also, this structure doesn't have any other methods, maybe it worth refactoring the code that relies on it to use `CeloBackend` and `GetBalanceERC20` directly then?

hbandura replied:
removed the struct

carterqw2 replied:
Any chance to combine `GetBalanceOf` and `GetBalanceERC20` into a single function and use it everywhere?

hbandura replied:
Since GetBalanceERC20 is used from the state_transition, I would dare not change it. And we need to check if the feeCurrency is for the native currency which now I removed GetBalanceOf, but added a CeloBackend.GetFeeBalance(account, feeCurrency), which is directly called from the pool.celoBackend of legacypool and blobpool.

carterqw2 commented on core/celo_evm.go:
```diff
@@ -16,7 +16,7 @@ import (
 
 // Returns the exchange rates for all gas currencies from CELO
-func getExchangeRates(caller *CeloBackend) (map[common.Address]*big.Rat, error) {
+func GetExchangeRates(caller bind.ContractCaller) (common.ExchangeRates, error) {
```
Perhaps this function could also move to `CeloBackend`?

hbandura replied:
Done

carterqw2 commented on core/txpool/blobpool/blobpool.go:
```diff
@@ -335,7 +335,7 @@ func New(config Config, chain BlockChain) *BlobPool {
 		spent:  make(map[common.Address]*uint256.Int),
 
-		feeCurrencyValidator: txpool.NewFeeCurrencyValidator(),
+		feeCurrencyValidator: txpool.NewFeeCurrencyValidator(chain.Config()),
```
I couldn't find where `feeCurrencyValidator` is used in blobpool.

hbandura replied:
in validateTx()

hbandura replied:
although that's not good enough reason as to have it as a member, I'll revert that

carterqw2 commented on core/txpool/celo_validation.go:
```diff
@@ -15,29 +16,24 @@ var NonWhitelistedFeeCurrencyError = errors.New("Fee currency given is not white
 // FeeCurrencyValidator validates currency whitelisted status at the specified
 // block number.
 type FeeCurrencyValidator interface {
```
Looks like the only thing this interface does is provides the `Balance` function. Looks like it could be another struct's responsibility. `CeloBackend` potentially, since `Balance` through a chain of proxy functions calls its `GetBalanceERC20` in the end anyway. Then we could get rid of this interface and corresponding struct completely.

hbandura replied:
Done

carterqw2 commented on core/txpool/celo_validation.go:
```diff
@@ -77,34 +48,13 @@ func (cvo *CeloValidationOptions) Accepts(txType uint8) bool {
-		if !fcv.IsWhitelisted(st, tx.FeeCurrency()) {
-			return NonWhitelistedFeeCurrencyError
-		}
+	if !common.IsCurrencyWhitelisted(rates, tx.FeeCurrency()) { // TODO: change to celoContext
```
Perhaps the comment is not relevant anymore?

hbandura replied:
Removed the comment

carterqw2 commented on core/txpool/validation.go:
```diff
@@ -241,9 +244,9 @@ func ValidateTransactionWithState(tx *types.Transaction, signer types.Signer, op
 	}
 	// Ensure the transactor has enough funds to cover the transaction costs
+	balance := opts.ExistingBalance(from, tx.FeeCurrency())
 	var (
```
Nit: it is worth either move `balance` back to the `var` section or move `cost` out as well, whichever makes a smaller diff.

hbandura replied:
changing it

carterqw2 commented on core/txpool/validation.go:
```diff
@@ -215,14 +215,17 @@ type ValidationOptionsWithState struct {
 	// L1CostFn is an optional extension, to validate L1 rollup costs of a tx
 	L1CostFn L1CostFunc
+
+	// ExistingBalance for a currency, to check for balanco to cover transaction costs.
```
A typo in `balanco`

hbandura replied:
fixed!

carterqw2 commented on core/txpool/blobpool/celo_blobpool.go:
```diff
@@ -0,0 +1,10 @@
+	"github.com/ethereum/go-ethereum/core/txpool"
+)
+
+func (b *BlobPool) currentRates() common.ExchangeRates {
```
In `celo_legacypool` it's `recalculateRates`, perhaps we should stay consistent.

hbandura replied:
changed it

carterqw2 commented on core/txpool/blobpool/blobpool.go:
```diff
@@ -1050,12 +1045,11 @@ func (p *BlobPool) validateTx(tx *types.Transaction) error {
 	}
 	// Ensure the transaction adheres to the stateful pool filters (nonce, balance)
-	stateOpts := &txpool.ValidationOptionsWithState{
+	opts := &txpool.ValidationOptionsWithState{
```
We can keep it `stateOpts` to reduce the diff.

hbandura replied:
you are right, I had mistaken here the variable name (it is 'opts' in the same place in the legacypool in upstream)

carterqw2 commented on core/txpool/legacypool/legacypool.go:
```diff
@@ -241,8 +241,8 @@ type LegacyPool struct {
-	// Celo
-	feeCurrencyValidator txpool.FeeCurrencyValidator
+	// Celo specific
+	currentRates common.ExchangeRates // current exchange rates for fee currencies
```
What if instead of `currentRates` we add an instance of `CeloBackend`? Then we would be able to get rid of `GetRatesForState` function and use `GetExchangeRates` of the backend directly. Additionally, we could combine `GetBalanceOf` and `GetBalanceERC20` function of the backend and call it directly and get rid of the `getBalances` and `recalculateRates` methods in `celo_legacypool` and perhaps `celo_blobpool` completely.

hbandura replied:
I don't discard it, but CeloBackend.GetExchangeRates still does many calls to contract code, I think it makes sense to have one instance per reset.
The instance of CeloBackend though I think it makes sense, and relcalculateRates use it to get the new ones. I'll make that change and let me know what you think.

carterqw2 commented on core/txpool/legacypool/list.go:
```diff
@@ -553,11 +544,13 @@ const (
+	p := &pricedList{
 		all: all,
 	}
+	p.floating.txComparator = txComparator
```
What if we change `txComparator` to be a pointer to the struct and store it on the priceList level? There are no priceHeap-specific functionality in the comparator, so both `floating` and `urgent` could share the same pointer as well.Then we could update rates in the comparator only once in that `SetBaseFeeAndRates` method below.

hbandura replied:
done

carterqw2 commented on core/txpool/legacypool/celo_list.go:
```diff
@@ -0,0 +1,170 @@
+}
+
+func (c *list) TotalCostFor(feeCurrency *common.Address) *big.Int {
+	if tc, ok := c.totalCost[getCurrencyKey(feeCurrency)]; ok {
```
Shouldn't we use `totalCostVar` here?

hbandura replied:
I intended this one to be just a retrieval function, but it won't change the internals.
totalCostVar is used when you do want to get the internal variable used to easily just add to it.

carterqw2 commented on core/txpool/legacypool/celo_list.go:
```diff
@@ -0,0 +1,170 @@
+}
+
+func (c *list) costCapFor(feeCurrency *common.Address) *big.Int {
+	if tc, ok := c.costCap[getCurrencyKey(feeCurrency)]; ok {
```
Similar here, shouldn't we have something like `CostCapVar` function that would transparently create items for missing keys and store them?

hbandura replied:
since costCap is not an accumulator, but only the max between(previous cap, max tx cost in list), I thought it looked better to just get the total cap, and, if a tx is accepted, just call to 'updateCap' which may or may not update it depending on the comparison

palango commented on core/celo_backend.go:
```diff
@@ -43,8 +48,66 @@ func (b *CeloBackend) CallContract(ctx context.Context, call ethereum.CallMsg, b
+	return balance, nil
+}
+
+func (b *CeloBackend) GetFeeBalance(account common.Address, feeCurrency *common.Address) *big.Int {
```
Please add a docstring here

hbandura replied:
done

palango commented on core/celo_backend.go:
```diff
@@ -43,8 +48,66 @@ func (b *CeloBackend) CallContract(ctx context.Context, call ethereum.CallMsg, b
+	}
+	balance, err := b.GetBalanceERC20(account, *feeCurrency)
+	if err != nil {
+		log.Error("Error while trying to get ERC20 balance:", "cause", err, "contract", feeCurrency.Hex, "account", account.Hex())
```
Looks like there are parenthesis missing for the `feeCurrency.Hex` call.

hbandura replied:
missed it, thanks

palango commented on core/celo_backend.go:
```diff
@@ -43,8 +48,66 @@ func (b *CeloBackend) CallContract(ctx context.Context, call ethereum.CallMsg, b
+	for _, tokenAddress := range whitelistedTokens {
+		numerator, denominator, err := oracle.MedianRate(&bind.CallOpts{}, tokenAddress)
+		if err != nil {
+			log.Error("Failed to get medianRate for gas currency!", "err", err, "tokenAddress", tokenAddress)
```
Add `.Hex()` call to log.

hbandura replied:
done

palango commented on core/celo_backend.go:
```diff
@@ -43,8 +48,66 @@ func (b *CeloBackend) CallContract(ctx context.Context, call ethereum.CallMsg, b
+			continue
+		}
+		if denominator.Sign() == 0 {
+			log.Error("Bad exchange rate for fee currency", "tokenAddress", tokenAddress, "numerator", numerator, "denominator", denominator)
```
Add `.Hex()` call to log.

hbandura replied:
done

palango commented on core/txpool/celo_rates.go:
```diff
@@ -0,0 +1,93 @@
+)
+
+func GetRatesForState(config *params.ChainConfig, state *state.StateDB) common.ExchangeRates {
+	stateCopy := state.Copy()
```
Do we actually need this copy as the exchange rates are only read?

hbandura replied:
forgot to remove this FN from the last refactor

palango commented on core/txpool/celo_rates.go:
```diff
@@ -0,0 +1,93 @@
+	unitRate = big.NewRat(1, 1)
+)
+
+func GetRatesForState(config *params.ChainConfig, state *state.StateDB) common.ExchangeRates {
```
Function seems unused

hbandura replied:
removed it

palango commented on core/txpool/celo_rates.go:
```diff
@@ -0,0 +1,93 @@
+	return rates
+}
+
+func ConvertCurrency(exchangeRates common.ExchangeRates, val1 *big.Int, currencyFrom *common.Address, currencyTo *common.Address) *big.Int {
```
Should we move this to the `ConvertCurrencyToGold` function?

hbandura replied:
What do you mean? move it to the same package/file ?

palango replied:
Yes, basically put all the conversion functions in a file/module.

hbandura replied:
I grouped exchanges and comparison into common/exchange

palango commented on core/txpool/celo_rates.go:
```diff
@@ -0,0 +1,93 @@
+	return toAmount
+}
+
+func getRate(exchangeRates common.ExchangeRates, currency *common.Address) (*big.Rat, error) {
```
We usually call this parameter `feeCurrency`, maybe use that for consistency.

hbandura replied:
done

palango commented on core/txpool/celo_rates.go:
```diff
@@ -0,0 +1,93 @@
+func ConvertCurrency(exchangeRates common.ExchangeRates, val1 *big.Int, currencyFrom *common.Address, currencyTo *common.Address) *big.Int {
+	goldAmount, err := contracts.ConvertCurrencyToGold(exchangeRates, val1, currencyFrom)
+	if err != nil {
+		log.Error("Error trying to convert from currency to gold.", "value", val1, "fromCurrency", currencyFrom)
```
Also add `Hex()` here for addresses.

hbandura replied:
done

palango commented on core/txpool/legacypool/legacypool.go:
```diff
@@ -1495,16 +1492,16 @@ func (pool *LegacyPool) promoteExecutables(accounts []common.Address) []*types.T
-		}
 		// Drop all transactions that are too costly (low balance or out of gas)
-		drops, _ := list.Filter(balance, gasLimit)
+		// var l1Cost *big.Int
```
Will this be done in this PR or a subsequent one?

hbandura replied:
Yes, I believe we need to discuss this after this PR is done

palango replied:
Please create a issue for tracking this then.

hbandura replied:
Created

palango commented on core/txpool/legacypool/list.go:
```diff
@@ -281,9 +280,11 @@ type list struct {
+	costCap map[common.Address]*big.Int // Price of the highest costing transaction per currency (reset only if exceeds balance)
+	gascap  uint64                      // Gas limit of the highest spending transaction (reset only if exceeds block limit)
+
+	// Celo additions for multi currency
```
`costCap` should be moved here as well if we keep this comment

hbandura replied:
I kept the new map in the same line as the previous costcap intentionally, since it's not an addition but a modification.

palango replied:
Fine for me, but then why change `totalcost` to `totalCost`?

hbandura replied:
same reason costcap changed to costCap, to make it more obvious on merge conflicts on the difference between our usage and upstream's.

palango replied:
That makes a lot of sense.

hbandura committed:
Move exchange comparisons and etc to common/exchange
